app/validators/uri_validator.rb

```diff
@@ -1,53 +1,102 @@
 # frozen_string_literal: true
 
 class UriValidator < ActiveModel::EachValidator
-  DEFAULT_ACCEPTED_SCHEMES = /^https?$/
-  DEFAULT_OPTIONAL_PARTS = %i[port].freeze
-  DEFAULT_FORBIDDEN_PARTS = %i[userinfo registry path opaque query fragment].freeze
-
-  def validate_each(record, attribute, value)
-    valid = begin
-              uri = URI.parse(value)
-              valid_scheme?(uri.scheme) && valid_host?(uri.host) && !forbidden_part?(record, uri)
-            rescue URI::InvalidURIError
-              false
-            end
-    record.errors.add(attribute, options[:message] || :invalid) unless valid
-    valid
+  def validate_each(record, attribute_name, attribute_value)
+    errors = UriPartsValidator.validate(attribute_value, self.class.options_for_record(options, record))
+    errors.each { |error| record.errors.add(attribute_name, error) }
+    errors.empty?
   end
 
-  def valid_scheme?(scheme)
-    accepted_scheme = options.fetch(:scheme, DEFAULT_ACCEPTED_SCHEMES)
-    case accepted_scheme
-    when Regexp
-      scheme =~ accepted_scheme
-    else
-      [*accepted_scheme].include? scheme
+  # This smells :reek:ManualDispatch
+  def self.options_for_record(options, record)
+    options.each_with_object({}) do |(options_name, options_value), opts|
+      opts[options_name] = options_value.respond_to?(:call) ? options_value.call(record) : options_value
     end
   end
 
-  def valid_host?(host)
-    return false if host.blank?
-    (host.size <= 255) && valid_host_labels?(host)
-  end
+  class UriPartsValidator
+    DEFAULT_ACCEPTED_SCHEMES = /^https?$/
+    DEFAULT_PERMISSIONS_OF_PARTS = {
+      port:     true,
+      userinfo: false,
+      registry: false,
+      path:     false,
+      opaque:   false,
+      query:    false,
+      fragment: false
+    }.freeze
+    MAX_HOST_SIZE = 255
+    MAX_LABEL_SIZE = 63
 
-  def valid_host_labels?(host)
-    host.split('.').map(&:size).none?(&63.method(:'<'))
```

Comment:
This didn't mean that each label could be up to 62-characters long. It called `63 < each_part_of_the_host.size`.

Reply:
Ah right 👍 sorry 🙏 i will fix that 😄